main/resource-access.bs

```diff
@@ -130,31 +125,150 @@ Note:
 Servers that wish to disable URI re-use may want to use the `410` status
 code.
 
+## Reading and Writing Resources ## {#read-write}
+
+Servers MUST respond with the `405` status code to requests using HTTP methods
+that are not supported by the target resource.
+[[Source](https://github.com/solid/specification/issues/117)]
+
 ### Resource type heuristics ### {#resource-type-heuristics}
 
 When creating new resources, servers can determine an effective request URI's
 type by examining the URI path ending ([[#uri-slash-semantics]]).
 
-Clients who want to assign a URI for their resource, MUST use PUT and PATCH
-requests. Servers MAY allow clients to suggest the URI for a resource created
+Clients who want to assign a URI to a resource, MUST use `PUT` and `PATCH`
+requests. Servers MAY allow clients to suggest the URI of a resource created
 through POST, using the HTTP `Slug` header as defined in [[!RFC5023]].
 
-Clients who want the server to determine the URI for their resource, MUST use
-the POST request.
+Clients who want the server to assign a URI of a resource, MUST use the `POST`
+request.
+
+[[Source](https://github.com/solid/specification/pull/160#issuecomment-636822687)].
+
+### Reading resources ### {#read}
+
+Servers MUST support the HTTP `GET`, `HEAD` and `OPTIONS` methods [[!RFC7231]]
+for clients to read resources or to determine communication options.
+[[Source](https://github.com/solid/specification/issues/39#issuecomment-538017667)]
+
+Servers MUST indicate their support for HTTP Methods by responding to HTTP
+`GET`, `HEAD` and `OPTIONS` requests on the resource's URL with the HTTP
```

Comment:
I am a little concerned about the inclusion of `OPTIONS` here. If `OPTIONS` requests need to include the appropriate `Allow` header, then consider the case where an RDF resource supports PATCH but a NonRDF resource doesn't. Because CORS needs to be able to send unauthenticated preflight requests to that URL, if `OPTIONS` responses distinguish between resource types to unauthenticated clients, then a malicious (unauthenticated) client can use this method to probe a server to find out information about what data is stored there. To include `OPTIONS` here, while also supporting CORS (which I assume is a given) would open up a server implementation to a variety of attacks and/or data leakages.

I would recommend simply removing `OPTIONS` from this list.

c.f. also https://www.mnot.net/blog/2012/10/29/NO_OPTIONS

Reply:
In the initial PR ( https://github.com/solid/specification/commit/51bbd733b2feaf2073ab6d8bf720332c67f9fb0c ) the `Allow` header was used in context of WAC. In https://github.com/solid/specification/pull/188#pullrequestreview-445094017 , Ruben and I agreed that the `Allow` header only indicates server's support - which was already stated in the spec: "Servers [..] indicate their support for HTTP Methods [..] on the resource's URL [..]"

Holding that perspective may mean that `Allow` does not need to distinguish between resource types to unauthenticated clients. Essentially `PATCH` is listed in `Allow` regardless of media types actually supported for a resource. Unless we are explicit about the methods to always list in `Allow` ie. equivalent to server requirements.

For `Accept-Patch` and `Accept-Post` headers in response to `OPTIONS` to be useful, it needs to distinguish between resource types. That would indeed expose information to unauthenticated clients. So, I agree to remove `OPTIONS` from the list, perhaps more specifically, the header should not be in the response to an authenticated request.

If `Accept-Patch` and `Accept-Post` values distinguish between resource types, would that mean that an agent needs to have `acl:Read` on the resource? Related issue: https://github.com/solid/specification/issues/43

Reply:
Revised: https://github.com/solid/specification/pull/188/commits/78507ffe0be507c8616073cd9dd487759efcdf77 . What do you think?

Reply:
@csarven that's great, thanks